### PR TITLE
[SPARK-33209][SS] Refactor unit test of stream-stream join in UnsupportedOperationsSuite

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

This PR is a followup from https://github.com/apache/spark/pull/30076 to refactor unit test of stream-stream join in `UnsupportedOperationsSuite`, where we had a lot of duplicated code for stream-stream join unit test, for each join type.

### Why are the changes needed?

Help reduce duplicated code and make it easier for developers to read and add code in the future.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing unit test in `UnsupportedOperationsSuite.scala` (pure refactoring).

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 1]
```diff
@@ -384,239 +384,166 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
...
+      streamStreamSupported = false,
+      expectedMsg = "is not supported in Update output mode")
+
+    // Full outer joins: only batch-batch is allowed
+    testBinaryOperationInStreamingPlan(
+      "full outer join",
+      _.join(_, joinType = FullOuter),
+      streamStreamSupported = false,
+      batchStreamSupported = false,
+      streamBatchSupported = false)
```
Comment (c21) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:415]: Note: this 4 tests are original from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L387-L415

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 2]
```diff
@@ -384,239 +384,166 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
...
+
+    // Left outer, left semi, left anti join: *-stream not allowed
+    Seq((LeftOuter, "LeftOuter join"), (LeftSemi, "LeftSemi join"), (LeftAnti, "Left anti join"))
+      .foreach { case (joinType, name) =>
+        testBinaryOperationInStreamingPlan(
```
Comment (c21) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:420]: Note:

* left outer join is from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L418 .

* left semi join is from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L488 .

* left anti join is from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L558 .

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 3]
```diff
@@ -384,239 +384,166 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
...
 
-  // Left outer joins: *-stream not allowed
-  testBinaryOperationInStreamingPlan(
-    "left outer join",
-    _.join(_, joinType = LeftOuter),
-    batchStreamSupported = false,
-    streamStreamSupported = false,
-    expectedMsg = "outer join")
+    // Right outer joins: stream-* not allowed
+    testBinaryOperationInStreamingPlan(
```
Comment (c21) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:429]: Note: this test is from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L566 .

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 4]
```diff
@@ -384,239 +384,166 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
...
+      "right outer join",
+      _.join(_, joinType = RightOuter),
+      streamBatchSupported = false,
+      streamStreamSupported = false,
+      expectedMsg = "outer join")
+
+    // Left outer, right outer, left semi joins
+    Seq(LeftOuter, RightOuter, LeftSemi).foreach { joinType =>
+      // Update mode not allowed
```
Comment (c21) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:438]: Note:

* left outer join is from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L425-L485 .

* right outer join is from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L573-L619 .

* left semi join is from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L495-L555 .

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 5]
```diff
@@ -384,239 +384,166 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
...
 
-  // Right outer joins: stream-* not allowed
-  testBinaryOperationInStreamingPlan(
-    "right outer join",
-    _.join(_, joinType = RightOuter),
-    streamBatchSupported = false,
-    streamStreamSupported = false,
-    expectedMsg = "outer join")
+    // stream-stream inner join doesn't emit late rows, whereas outer joins could
+    Seq((Inner, false), (LeftOuter, true), (RightOuter, true)).map {
```
Comment (c21) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:501]: Note: this test is from https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L744 .

**Comment by c21** [comment]
cc @cloud-fan , @HeartSaVioR , @xuanyuanking , @sameeragarwal if you guys have time to take a look, thanks.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 6]
```diff
@@ -384,239 +384,166 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
...
-
-  // Full outer joins: only batch-batch is allowed
-  testBinaryOperationInStreamingPlan(
-    "full outer join",
-    _.join(_, joinType = FullOuter),
-    streamStreamSupported = false,
-    batchStreamSupported = false,
-    streamBatchSupported = false)
+  // stream-stream join
+  {
```
Comment (HeartSaVioR) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:388]: Let's remove the bracket for now, as it produces unnecessary changes. If we think it's better to have it to separate properly, we can add it after verifying all changes are OK.

Comment (c21) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:388]: Similar to existing unit tests in this file, i.e. `streaming aggregation`, `FlatMapGroupsWithState`, `deduplicate`, I add bracket to group `stream-stream join` here, I think this is part of refactoring, as we move all relevant tests together, and it looks better. But it's just a pretty minor thing after all, and I am fine to remove it if people think that way.

Comment (HeartSaVioR) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:388]: Without this bracket you can reduce many line changes which are hurdles to review. That can be added later (after reviewing) as I mentioned in my previous comment. Makes sense?

Comment (c21) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala]: cool, I got your point now, updated accordingly, thanks.

**Comment by cloud-fan** [comment]
looks fine

**Comment by c21** [comment]
@HeartSaVioR - could you also help take a look when you have time? Thanks.

**Comment by HeartSaVioR** [comment]
Hey sorry. Please give me to some more time to start looking into, as here's Sat. already. I'd probably take this in early next week. Thanks for understanding.

**Comment by c21** [comment]
Thanks @HeartSaVioR for heads up, please take your time. Thanks.

**Review by HeartSaVioR — APPROVED** [review]
+1

I'll wait for a day to see the chance other reviewer could participate later and find my missing. If there's no further comment, I'll merge this tomorrow.

Also cc. to @gaborgsomogyi

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 7]
```diff
@@ -414,209 +412,135 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
     batchStreamSupported = false,
     streamBatchSupported = false)
 
-  // Left outer joins: *-stream not allowed
+  // Left outer, left semi, left anti join: *-stream not allowed
+  Seq((LeftOuter, "LeftOuter join"), (LeftSemi, "LeftSemi join"), (LeftAnti, "Left anti join"))
```
Comment (xuanyuanking) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:416]: super nit, let's keep naming style here? `LeftOut join` ... `LeftAnti join` or `left outer join` ... `left anti join`.

**Review by xuanyuanking — APPROVED** [review]
LGTM

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala (hunk 7 above)
Comment (c21) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:416]: @xuanyuanking - wondering what's the naming style we are trying to following here? TBH I would be surprised that someone fails to understand comment of `left outer`, if I omit `join` here. Similar comments happened several times in other files e.g. in [StreamingSymmetricHashJoinExec](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala#L325) . But anyway I feel it's too nit, if you think it's worth the change, I can do it as well.

Comment (xuanyuanking) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:416]: I mean at least be consistent for the `name` param here:
`LeftOuter join` `LeftSemi join` and `LeftAnti join`
or
`left outer join` `left simi join` and `left anti join`

Yes, it's too nit and not worth a single commit if there's no further comment.

Comment (HeartSaVioR) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:416]: This is nit but super easy to fix, so probably better to fix now once we find it. It'd be appreciated if you can fix this - won't drag the review process as @xuanyuanking already provided LGTM and this is probably going to be a final review comment.

Comment (HeartSaVioR) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:416]: @c21 
No his suggestion is not about the comment. It's in `name` part, `Left anti join` -> `LeftAnti join`, as for other joins you don't add a space.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 8]
```diff
@@ -414,209 +412,135 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
     batchStreamSupported = false,
     streamBatchSupported = false)
 
-  // Left outer joins: *-stream not allowed
+  // Left outer join, left semi join, left anti join: *-stream not allowed
+  Seq((LeftOuter, "LeftOuter join"), (LeftSemi, "LeftSemi join"), (LeftAnti, "Left anti join"))
```
Comment (HeartSaVioR) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:416]:
```suggestion
  Seq((LeftOuter, "LeftOuter join"), (LeftSemi, "LeftSemi join"), (LeftAnti, "LeftAnti join"))
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala (hunk 7 above)
Comment (c21) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala]: Sorry that I misunderstood in the first place. Changed to `LeftAnti join` and https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala#L307 as well.

**Review by HeartSaVioR — APPROVED** [review]
+1 again, pending build.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Comment by c21** [comment]
Thanks @HeartSaVioR , @cloud-fan and @xuanyuanking for review!
